### PR TITLE
support group based authorization without 'Read directory data' permission

### PR DESCRIPTION
**Pull request by amoosbr**
Hi,

this pull-request introduces support for group based Azure AD authentication and authorization without needing 'Read directory data' permissions.
Instead, Azure Active directory is configured to send the user groups as part of the ID token. This setting doesn't need admin rights.

It was a quick test yesterday and worked surprisingly perfectly for me. But I wasn't able to verify, that the 'Read directory data' group support is still working fine.
I updated the README accordingly and hope it helps others to verify this change is working for them as well.

### CODE REVIEW

**Comment by amoosbr** [comment]
with this MR it might be possible to close #8

File: src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java [hunk 1]
```diff
@@ -108,6 +126,7 @@ public boolean equals(Object o) {
         if (familyName != null ? !familyName.equals(that.familyName) : that.familyName != null) return false;
         if (uniqueName != null ? !uniqueName.equals(that.uniqueName) : that.uniqueName != null) return false;
         if (tenantID != null ? !tenantID.equals(that.tenantID) : that.tenantID != null) return false;
+        if (groupOIDs != null ? !groupOIDs.equals(that.groupOIDs) : that.groupOIDs != null) return false;
```
Comment (gavinfish) [inline src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java]: Is it possible that the `groupOIDs` has same elements with different orders?

File: src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java [hunk 2]
```diff
@@ -84,11 +93,20 @@ public static AzureAdUser createFromJwt(JwtClaims claims) {
     }
 
     public void setAuthorities(Collection<ActiveDirectoryGroup> groups) {
-        GrantedAuthority[] newAuthorities = new GrantedAuthority[groups.size() * 2 + 2];
+        GrantedAuthority[] newAuthorities;
         int i = 0;
-        for (ActiveDirectoryGroup group : groups) {
-            newAuthorities[i++] = new AzureAdGroup(group.id(), group.name());
-            newAuthorities[i++] = new GrantedAuthorityImpl(group.id());
+        if (!groups.isEmpty()) {
```
Comment (gavinfish) [inline src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java]: `groups` may be null.

File: src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java (hunk 1 above)
Comment (amoosbr) [reply src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java]: You're right with the comment. I'm not sure the lists are always in the same sequence. Provided a fix for this.

File: src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java (hunk 2 above)
Comment (amoosbr) [reply src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java]: I didn't add a null check for `groups`, as the previous implementation also assumed `groups` to never be null.  See first line in old code: `GrantedAuthority[] newAuthorities = new GrantedAuthority[groups.size() * 2 + 2];`.
But if that is not the case, I can also add a null check at the beginning of the method.
Just let me know. And thanks for the quick review

Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java]: You are right. I should fix the potential risk by changing the code at:

https://github.com/jenkinsci/azure-ad-plugin/blob/f39cb0314f6c6789cd19647c5d159645e29f965d/src/main/java/com/microsoft/jenkins/azuread/AzureCachePool.java#L65-L68

**Pull request merged**